docs/adr/0008-wrapper-code.md

```diff
@@ -0,0 +1,91 @@
+
+We should decide how much effort, if any, we want to invest into getting rid of the wrapper code now that it's technically possible.
+We should also decide if there are places where wrapper code makes sense and we don't want to replace it, even ignoring the time investment.
+
+### Possible changes
+
+This section explores some possibilities for removing wrapper code.
+Deciding on any particular possibility listed here is out of scope for the ADR, the decision is if we should invest our time investigating some of these.
+
+#### Async
+
+One of the main reasons for our current wrapper code is to wrap our sync API and present an async interface.
+ADR-0009 lays out a couple of alternatives to this.
+
+#### Feature flags for breaking changes
+
+Another main reason for our current wrapper code is to mitigate breaking API changes.
+The wrapper layer allows us to make breaking changes at the Rust level, but keep the same API at the wrapper layer.
+
+An alternative to this would be to use Rust feature flags to manage breaking changes.
+Any breaking change, or large change in general, would be behind a feature flag.
+We would wait to enable the feature flag on the megazord until consumer application was ready to take the change.
+Maybe we could have a transition period where we built two megazords, for example `megazord` and `megazord-with-logins-local-encrytion` and the consumer app could pick between the two.
+This would simplify the consumer PRs since they could run CI normally.
+
+Some potential uses of of features flags would be:
+
+- **Cosmetic changes**. If we want to rename a field/function name, we could put that rename behind a feature flag.
+- **Architectural changes**. For bigger changes, like the logins local encryption rework, we could maintain two pieces of code at once. For example, copy db.rs to db/old.rs and db/new.rs. Create db/mod.rs which runs `use old::*` or `use new::*` depending on the feature flag. Then we do our work in db/new.rs.
+
+Maintaining many feature flags at once would require significant effort, so we should aim to get our consumers to use the new feature flag soon after our work is complete.
+
+#### UniFFI-supported interfaces
+
+One last reason for wrapper code is to present idiomatic interfaces to our Rust code -- especially for callback interfaces.
+For example, it's possible to define a UniFFI callback interface for notifications, but the FxA wrapper code uses the `NotificationCenter` on Swift which is not supported by UniFFI.
+If we wanted to remove all wrapper code we would need to commit to only using interfaces that UniFFI could support.
+
+## Decision Drivers
+
+## Considered Options
+
+* **(A) Keep using our current strategy**
+Don't invest time into removing wrapper code.
+
+* **(B) Remove all wrapper code**
+Invest time into removing wrapper code with the intention of removing all of it.
+
+* **(C) Remove most wrapper code, keep an additive wrapper layer**
```

Comment:
I agree with other comments that suggest we need to analyze our current wrapper code to see how much of it we can remove and that the result will inform our future strategy. I for one am assuming that we won't be able to remove all of it so we will end up at option C. We also might want to have a one-to-one wrapper to consumer strategy so that we can put the more platform-specific, opinionated code there. But in order to reduce our API surface I think we should do what we can to reduce the wrapper code we have.